farming/logics/traits/master_chef/farming.rs

```diff
@@ -0,0 +1,482 @@
+            }
+            pool.last_reward_block = current_block;
+            self.data::<Data>().pool_info.insert(pool_id, &pool);
+
+            self._emit_log_update_pool_event(
+                pool_id,
+                pool.last_reward_block,
+                lp_supply,
+                pool.acc_arsw_per_share,
+            );
+        }
+        Ok(())
+    }
+
+    fn _calculate_additional_acc_arsw_per_share(
+        &mut self,
+        pool_info: &Pool,
+        current_block: u32,
+        lp_supply: Balance,
+    ) -> Result<Balance, FarmingError> {
+        if lp_supply == 0 {
+            return Err(FarmingError::LpSupplyIsZero)
+        }
+        let last_reward_block_period = self._get_period(pool_info.last_reward_block)?;
+        let current_period = self._get_period(Self::env().block_number())?;
+
+        let mut arsw_reward: Balance = 0;
+        let mut last_block = pool_info.last_reward_block;
+        let mut period = last_reward_block_period;
+        while period <= current_period {
+            if period > MAX_PERIOD {
+                break
+            }
+            let total_alloc_point: u32 = self.data::<Data>().total_alloc_point;
+            if current_block <= self._period_max(period)? {
+                arsw_reward = arsw_reward
+                    .checked_add(
+                        (current_block as u128)
+                            .checked_sub(last_block as u128)
+                            .ok_or(FarmingError::SubUnderflow4)?
+                            .checked_mul(self._arsw_per_block(period)?)
+                            .ok_or(FarmingError::MulOverflow3)?
+                            .checked_mul(pool_info.alloc_point as u128)
+                            .ok_or(FarmingError::MulOverflow4)?
+                            .into(),
+                    )
+                    .ok_or(FarmingError::AddOverflow6)?
+                    .checked_div(total_alloc_point.into())
+                    .ok_or(FarmingError::DivByZero1)?
+            } else {
+                arsw_reward = arsw_reward
+                    .checked_add(
+                        (self._period_max(period)? as u128)
+                            .checked_sub(last_block.into())
+                            .ok_or(FarmingError::SubUnderflow5)?
+                            .checked_mul(self._arsw_per_block(period)? as u128)
+                            .ok_or(FarmingError::MulOverflow5)?
+                            .checked_mul(
+                                pool_info
+                                    .alloc_point
+                                    .checked_div(total_alloc_point)
+                                    .ok_or(FarmingError::DivByZero2)?
+                                    .into(),
+                            )
+                            .ok_or(FarmingError::MulOverflow6)?
+                            .into(),
+                    )
+                    .ok_or(FarmingError::AddOverflow7)?;
```

Comment:
Same for here. In this case division should be done to addend; not to the alloc point otherwise the right addend will result to 0.